src/mod_ping.erl

```diff
@@ -1,130 +1,101 @@
-    %% resource by accident. If this happens the server ends up in a
-    %% hot loop of iq-error replies.
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    gen_server:cast(Proc, {start_ping, JID});
-start_ping(_Host, _JID) -> ok.
+handle_info(init, HandlerState, C2SState) ->
+    {add_ping_timer(HandlerState, C2SState), C2SState};
+handle_info(send_ping, HandlerState, C2SState) ->
+    JID = ejabberd_c2s_state:jid(C2SState),
+    Server = ejabberd_c2s_state:server(C2SState),
+    route_ping_iq(JID, Server),
+    {add_ping_timer(HandlerState, C2SState), C2SState};
+handle_info(timeout, HandlerState, C2SState) ->
+    JID = ejabberd_c2s_state:jid(C2SState),
+    Server = ejabberd_c2s_state:server(C2SState),
+    ejabberd_hooks:run(user_ping_timeout, Server, [JID]),
+    case gen_mod:get_module_opt(Server, ?MODULE, timeout_action, none) of
+        kill -> ejabberd_c2s:stop(self());
+        _ -> ok
+    end,
+    {HandlerState, C2SState}.
 
-stop_ping(Host, JID) ->
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    gen_server:cast(Proc, {stop_ping, JID}).
+add_ping_timer(HandlerState, C2SState) ->
+    cancel_timer(HandlerState),
+    Server = ejabberd_c2s_state:server(C2SState),
+    PingInterval = gen_mod:get_module_opt(Server, ?MODULE, ping_interval, ?DEFAULT_PING_INTERVAL),
+    erlang:send_after(PingInterval * 1000, self(), {mod_ping, send_ping}).
 
-%%====================================================================
-%% gen_mod callbacks
-%%====================================================================
-start(Host, Opts) ->
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    PingSpec = {Proc, {?MODULE, start_link, [Host, Opts]},
-                transient, 2000, worker, [?MODULE]},
-    supervisor:start_child(?SUPERVISOR, PingSpec).
+route_ping_iq(JID, Server) ->
+    PingReqTimeout = gen_mod:get_module_opt(Server, ?MODULE, ping_req_timeout,
+                                            ?DEFAULT_PING_REQ_TIMEOUT),
+    IQ = #iq{type = get,
+             sub_el = [#xmlel{name = <<"ping">>,
+                              attrs = [{<<"xmlns">>, ?NS_PING}]}]},
+    Pid = self(),
+    F = fun(_, _, _, timeout) ->
+               Pid ! {mod_ping, timeout};
+           (_From, _To, Acc, _Response) ->
+                Acc
+        end,
+    From = jid:make(<<"">>, Server, <<"">>),
+    Acc = mongoose_acc:new(#{ location => ?LOCATION,
+                              lserver => Server,
+                              from_jid => From,
+                              to_jid => JID,
+                              element => jlib:iq_to_xml(IQ) }),
+    ejabberd_local:route_iq(From, JID, Acc, IQ, F, PingReqTimeout).
 
-stop(Host) ->
-    Proc = gen_mod:get_module_proc(Host, ?MODULE),
-    Pid = erlang:whereis(Proc),
-    gen_server:call(Proc, stop),
-    wait_for_process_to_stop(Pid),
-    supervisor:delete_child(?SUPERVISOR, Proc).
+cancel_timer(undefined) ->
+    do_nothing;
+cancel_timer(TRef) ->
+    erlang:cancel_timer(TRef).
```

Comment:
This function is redundant, let's remove it:

> If Result is false, a timer corresponding to `TimerRef` could not be found. This can be either because the timer had expired, already had been canceled, or because TimerRef never corresponded to a timer.
-- http://erlang.org/doc/man/erlang.html#cancel_timer-2

```erlang
1> R = make_ref().
#Ref<0.2521718886.3805282306.126943>
2> erlang:cancel_timer(R).
false
```

Reply:
@erszcz how about this:
```erlang
erlang:cancel_timer(undefined).
** exception error: bad argument
     in function  erlang:cancel_timer/1
        called as erlang:cancel_timer(undefined)
```

I think we still need to handle the case when `TRef` is `undefined` unless we change the code and allow only references as value for `TRef`.

Reply:
@michalwski Indeed, I didn't catch all the call sites :| My bad.

Reply:
Now, considering your other [comment](https://github.com/esl/MongooseIM/pull/2531/files#r392120978) regarding avoiding `undefined`, in this case we would need to come up with other data for the initial state of the `c2s_info_handler` which doesn't crash the `erlang:cancel_timer`.